<a href="https://colab.research.google.com/github/Hiramb-Yadav/Minor/blob/main/Universities_EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# *REWORK ON UNIVERSITIES DATASET*

## Problem Statement
Identifaction of the top 10 countries most sutable for receving fundings from **Multilateral Organizations** ( World Bank, UNESCO, IMF )

### Approach
1. Identifying countries with low average rank but, have some very high ranking universities.    *(Creating a Prime_list of 20 countries)*
2. Filtering and removing developed countries from the list.    *( assuming they are developin countries) (Cutting the Prime_list to 10)*
3. Filtering countries with political stability.    *(Cutting the Prime_list to 5)*
4. Adding countries with average or good ranking which are recovering form a war of a Crisis.    *(Adding 2 to the Prime_List )*

In [ ]:
# Importing Libaries
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# Importing data set (csv)
df = pd.read_csv('/content/drive/MyDrive/Complete_Projects/Top Hidden Universities/top universities.csv')

# Aligning the output format
pd.set_option('display.width',200) # Increasing the width
pd.set_option('display.max_columns',None)

print(df.head())

                                          University  Country     City  Global Rank
0  Université des Sciences et de la Technologie H...  Algeria  Algiers         3677
1           Université Benyoucef Benkhedda d’Alger 1  Algeria  Algiers         4919
2          École Nationale Supérieure d’Informatique  Algeria  Algiers         6297
3                      École Nationale Polytechnique  Algeria  Algiers         6411
4       Université Abou el Kacem Saâdallah d’Alger 2  Algeria  Algiers         6421


# Data Cleaning

### Understanding the data set

In [ ]:
print(df)

                                              University  Country     City  Global Rank
0      Université des Sciences et de la Technologie H...  Algeria  Algiers         3677
1               Université Benyoucef Benkhedda d’Alger 1  Algeria  Algiers         4919
2              École Nationale Supérieure d’Informatique  Algeria  Algiers         6297
3                          École Nationale Polytechnique  Algeria  Algiers         6411
4           Université Abou el Kacem Saâdallah d’Alger 2  Algeria  Algiers         6421
...                                                  ...      ...      ...          ...
12809  Facultad de Teologa del Uruguay Mons. Mariano ...  Uruguay  Uruguay        11518
12810                      Politécnico de Punta del Este  Uruguay  Uruguay        11549
12811  Instituto Universitario de Docencia, Investiga...  Uruguay  Uruguay        11572
12812          Instituto Metodista Universitario Crandon  Uruguay  Uruguay        11630
12813  Instituto Universitario d

In [ ]:
print(df.describe())

        Global Rank
count  12814.000000
mean    6377.158499
std     3670.400228
min        1.000000
25%     3200.250000
50%     6384.500000
75%     9552.750000
max    12725.000000


In [ ]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12814 entries, 0 to 12813
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   University   12814 non-null  object
 1   Country      12814 non-null  object
 2   City         12814 non-null  object
 3   Global Rank  12814 non-null  int64 
dtypes: int64(1), object(3)
memory usage: 400.6+ KB
None


In [ ]:
# Checking the top ranks
sorted_df = df.sort_values(by = 'Global Rank', ascending = True)
print(sorted_df.head())

                                  University         Country           City  Global Rank
10760  Massachusetts Institute of Technology             Usa  Massachusetts            1
10761                     Harvard University             Usa  Massachusetts            2
12777                             Total Area  United-Kingdom              1            2
10154                    Stanford University             Usa     California            3
11073                     Cornell University             Usa       New York            4


In [ ]:
# Making the default data set sorted by Global Rank
df = sorted_df
print(df.head())

                                  University         Country           City  Global Rank
10760  Massachusetts Institute of Technology             Usa  Massachusetts            1
10761                     Harvard University             Usa  Massachusetts            2
12777                             Total Area  United-Kingdom              1            2
10154                    Stanford University             Usa     California            3
11073                     Cornell University             Usa       New York            4


In [ ]:
# Checking the number of duplicate entries in top 10 ranks
print(df.head(15))

                                        University         Country                City  Global Rank
10760        Massachusetts Institute of Technology             Usa       Massachusetts            1
10761                           Harvard University             Usa       Massachusetts            2
12777                                   Total Area  United-Kingdom                   1            2
10154                          Stanford University             Usa          California            3
11073                           Cornell University             Usa            New York            4
10155           University of California, Berkeley             Usa          California            5
10839                       University of Michigan             Usa            Michigan            6
11766                     University of Washington             Usa          Washington            7
11074  Columbia University in the City of New York             Usa            New York            8


In [ ]:
# Checking if there is duplicate entries of universities
df['University'].duplicated().sum()

np.int64(185)

In [ ]:
# Trying to get more intutive sense
df[df['University'].duplicated(keep=False)].sort_values(by='University')

,University,Country,City,Global Rank
5975,Ajou University,Korea,Gyeonggi,2345
5285,Ajou University,Korea,Gyeonggi,2345
5331,Andong National University,Korea,North Gyeongsang,6469
6021,Andong National University,Korea,North Gyeongsang,6469
5396,Baekseok University,Korea,South Chungcheong,6376
...,...,...,...,...
6029,Youngnam Theological University and Seminary,Korea,North Gyeongsang,11116
5409,Youngsan University,Korea,South Gyeongsang,6931
6099,Youngsan University,Korea,South Gyeongsang,6931
6113,Youngsan University of Son Studies,Korea,South Jeolla,11683


### #Observations
1. Most important variable = Global Rank
2. Total entries = 12814
3. Duplicate Ranks = True (I am avoiding the concept of dense rank to complete this project befor deadline.)
4. Duplicate University = True


### Removing duplicate entries

In [ ]:
# Deleting duplicated entries in University column
df = df.drop_duplicates(subset=['University'], keep='first')

# Deleting duplicated entries in Global Rank coumn
df= df.drop_duplicates(subset=['Global Rank'], keep='first')

# Verifying
df[df['University'].duplicated(keep=False)].sort_values(by='University')

,University,Country,City,Global Rank


In [ ]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
Index: 12626 entries, 10760 to 447
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   University   12626 non-null  object
 1   Country      12626 non-null  object
 2   City         12626 non-null  object
 3   Global Rank  12626 non-null  int64 
dtypes: int64(1), object(3)
memory usage: 493.2+ KB
None


# Finding Insights

## Milestone 1

In [ ]:
# Identifying countries which have low average rank but, have some very high ranking universities.

# Creating a new data set "df_countries "with columns namely - Country, Avg Rank, no. of universities in top 100.

# PHASE I
df_avg_rank = df.groupby('Country')['Global Rank'].mean().reset_index()   # I don't understand the working of this code
df_avg_rank.rename(columns={'Global Rank' : 'Avg Rank'}, inplace=True)    # I don't know where to use which column

# PHASE II
df_top_500 = df[df['Global Rank'] <= 500]
df_top_universities = df_top_500.groupby('Country')['University'].count().reset_index()
df_top_universities = df_top_universities.rename(columns={'University' : 'no. of top U'})

# PHASE III (merging)
df_countries = df_avg_rank.merge(df_top_universities, on='Country', how = 'left')

# Filling the missing values
df_countries['no. of top U'] = df_countries['no. of top U'].fillna(0)

# Sorting
df_countries = df_countries.sort_values(by='no. of top U', ascending=False)

print(df_countries.head(40))

            Country     Avg Rank  no. of top U
161             Usa  2746.225189         195.0
53          Germany  3559.079890          43.0
159  United-Kingdom  1583.716867          43.0
5         Australia   618.210526          21.0
25           Canada  2543.171717          21.0
141           Spain  1752.318841          18.0
71            Japan  5122.759547          13.0
105     Netherlands  2573.260000          11.0
49           France  3247.209424          11.0
69            Italy  2095.536585          10.0
18           Brazil  3815.297980           7.0
145          Sweden  3232.040000           7.0
13          Belgium  5016.196429           6.0
146     Switzerland  2120.875000           6.0
96           Mexico  9345.696691           6.0
30            China  6182.153846           6.0
139    South-Africa  3021.818182           4.0
60        Hong-Kong  1773.272727           4.0
68           Israel  4296.156250           4.0
124          Russia  5155.261838           4.0
6           A

In [ ]:
# Finding the University with having a value >= 0 in the column "no. of Top U" and having the worst Avg Rank
df_not_0 = df_countries[df_countries["no. of top U"] > 0]

# Now filtring down countries with worst rank
df_worst_R = df_not_0.sort_values(by='Avg Rank', ascending = False)

print(df_worst_R.head(40))

          Country     Avg Rank  no. of top U
96         Mexico  9345.696691           6.0
33     Costa-Rica  8789.207547           1.0
136      Slovenia  8783.166667           1.0
64      Indonesia  8721.431304           2.0
36         Cyprus  8423.591837           1.0
117          Peru  7236.676471           2.0
75          Korea  6911.309783           1.0
120      Portugal  6574.457831           3.0
31       Colombia  6396.180328           2.0
30          China  6182.153846           6.0
151      Thailand  5839.880000           1.0
124        Russia  5155.261838           4.0
71          Japan  5122.759547          13.0
4       Argentina  5082.421053           4.0
61        Hungary  5029.812500           2.0
13        Belgium  5016.196429           6.0
129  Saudi-Arabia  4760.827586           1.0
38        Denmark  4751.320000           4.0
41        Ecuador  4703.081633           1.0
148        Taiwan  4665.617647           1.0
119        Poland  4580.821429           3.0
37        

In [ ]:
print(df_countries.query("Country == 'India'"))

   Country     Avg Rank  no. of top U
63   India  8059.862857           0.0


In [ ]:
# Creating Prime List
prime_list_1 = df_worst_R.head(20)

print(prime_list_1)

          Country     Avg Rank  no. of top U
96         Mexico  9345.696691           6.0
33     Costa-Rica  8789.207547           1.0
136      Slovenia  8783.166667           1.0
64      Indonesia  8721.431304           2.0
36         Cyprus  8423.591837           1.0
117          Peru  7236.676471           2.0
75          Korea  6911.309783           1.0
120      Portugal  6574.457831           3.0
31       Colombia  6396.180328           2.0
30          China  6182.153846           6.0
151      Thailand  5839.880000           1.0
124        Russia  5155.261838           4.0
71          Japan  5122.759547          13.0
4       Argentina  5082.421053           4.0
61        Hungary  5029.812500           2.0
13        Belgium  5016.196429           6.0
129  Saudi-Arabia  4760.827586           1.0
38        Denmark  4751.320000           4.0
41        Ecuador  4703.081633           1.0
148        Taiwan  4665.617647           1.0


## Milestone 2

In [ ]:
# Printing just the name of countries prime_list_1
print(prime_list_1['Country'].to_list())

['Mexico', 'Costa-Rica', 'Slovenia', 'Indonesia', 'Cyprus', 'Peru', 'Korea', 'Portugal', 'Colombia', 'China', 'Thailand', 'Russia', 'Japan', 'Argentina', 'Hungary', 'Belgium', 'Saudi-Arabia', 'Denmark', 'Ecuador', 'Taiwan']


In [ ]:
# Deleting the following devloped countries from prime_list_1, Slovenia; Korea; Portugal; Japan;  Hungary; Belgium; Denmark; Taiwan.
developed_countries = ["Slovenia", "Korea", "Portugal", "Japan", "Hungary", "Belgium", "Denmark", "Taiwan"]

prime_list_1 = prime_list_1[~prime_list_1['Country'].isin(developed_countries)]

print(prime_list_1)

          Country     Avg Rank  no. of top U
96         Mexico  9345.696691           6.0
33     Costa-Rica  8789.207547           1.0
64      Indonesia  8721.431304           2.0
36         Cyprus  8423.591837           1.0
117          Peru  7236.676471           2.0
31       Colombia  6396.180328           2.0
30          China  6182.153846           6.0
151      Thailand  5839.880000           1.0
124        Russia  5155.261838           4.0
4       Argentina  5082.421053           4.0
129  Saudi-Arabia  4760.827586           1.0
41        Ecuador  4703.081633           1.0


In [ ]:
# Checking number of rows in prime_list_1
print(len(prime_list_1))

12


In [ ]:
# Deleting Saudi-Arabia and Cuprus from prime_list_1

prime_list_1 = prime_list_1[prime_list_1['Country'] != 'Saudi-Arabia']
prime_list_1 = prime_list_1[prime_list_1['Country'] != 'Cyprus']

prime_list_2 = prime_list_1

print(prime_list_2)

        Country     Avg Rank  no. of top U
96       Mexico  9345.696691           6.0
33   Costa-Rica  8789.207547           1.0
64    Indonesia  8721.431304           2.0
117        Peru  7236.676471           2.0
31     Colombia  6396.180328           2.0
30        China  6182.153846           6.0
151    Thailand  5839.880000           1.0
124      Russia  5155.261838           4.0
4     Argentina  5082.421053           4.0
41      Ecuador  4703.081633           1.0


## Milestone 3

In [ ]:
print(prime_list_2['Country'].to_list())

['Mexico', 'Costa-Rica', 'Indonesia', 'Peru', 'Colombia', 'China', 'Thailand', 'Russia', 'Argentina', 'Ecuador']


In [ ]:
prime_list_2 = prime_list_2[prime_list_2['Country'] != 'Ecuador']
prime_list_2 = prime_list_2[prime_list_2['Country'] != 'Argentina']
prime_list_2 = prime_list_2[prime_list_2['Country'] != 'Mexico']
prime_list_2 = prime_list_2[prime_list_2['Country'] != 'Colombia']
prime_list_2 = prime_list_2[prime_list_2['Country'] != 'Peru']

prime_list_3 = prime_list_2

print(prime_list_3)

        Country     Avg Rank  no. of top U
33   Costa-Rica  8789.207547           1.0
64    Indonesia  8721.431304           2.0
30        China  6182.153846           6.0
151    Thailand  5839.880000           1.0
124      Russia  5155.261838           4.0


## Milestone 4

In [ ]:
# Defining the set of recovering countries
recovering_countries = {'Ukraine', 'Iraq', 'Sierra Leone', 'Rwanda', 'Bosnia and Herzegovina'}

# Filtering universities in recovering countries with Global Rank < 2000
recovering_df = df[(df['Country'].isin(recovering_countries)) & (df['Global Rank'] < 5000)]

# Displaying the result
print(recovering_df)

                                             University  Country                    City  Global Rank
8961  National Technical University of Ukraine Kyiv ...  Ukraine             Kiev Oblast         1088
9003               Lviv Polytechnic National University  Ukraine             Lviv Oblast         1417
8928          V. N. Karazin Kharkiv National University  Ukraine          Kharkiv Oblast         1792
8962       Taras Shevchenko National University of Kyiv  Ukraine             Kiev Oblast         1896
9004            Ivan Franko National University of Lviv  Ukraine             Lviv Oblast         1945
8963         National University of Kyiv-Mohyla Academy  Ukraine             Kiev Oblast         2028
9045                              Sumy State University  Ukraine             Sumy Oblast         2073
8964                   Borys Grinchenko Kyiv University  Ukraine             Kiev Oblast         2931
8965         National Pedagogical Dragomanov University  Ukraine             Kiev 

## Final

In [ ]:
# Creating final_list
final_list = {'Cost-Rica', 'Indonesia', 'China', 'Thailand', 'Russia', 'Ukraine', 'Rwanda'}

# Filtering the dataset for only the countries in final_list
filtered_df = df[df['Country'].isin(final_list)]

# Sorting by 'Country' and 'Global Rank' (ascending order)
filtered_df = filtered_df.sort_values(by=['Country', 'Global Rank'], ascending=[True, True])

# Getting the top 5 universities in each country
top_5_per_country = filtered_df.groupby('Country').head(5)

# Display the result
print(top_5_per_country)

                                             University    Country               City  Global Rank
1354                                Tsinghua University      China              Beijg           44
1355                                  Peking University      China              Beijg          181
2219                                Zhejiang University      China           Zhejiang          239
2061                      Shanghai Jiao Tong University      China           Shanghai          241
1800                                 Nanjing University      China            Jiangsu          430
3634                              Universitas Indonesia  Indonesia          West Java          390
3733                            Universitas Gadjah Mada  Indonesia      Yogyakarta SR          414
3303                              Universitas Brawijaya  Indonesia          East Java          644
3635                         Institut Teknologi Bandung  Indonesia          West Java          693
3304      